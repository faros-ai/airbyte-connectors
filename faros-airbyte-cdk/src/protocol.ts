import {Dictionary} from 'ts-essentials';
import {VError} from 'verror';

import {wrapApiError} from './errors';

export enum AirbyteLogLevel {
  FATAL = 'FATAL',
  ERROR = 'ERROR',
  WARN = 'WARN',
  INFO = 'INFO',
  DEBUG = 'DEBUG',
  TRACE = 'TRACE',
}

export function AirbyteLogLevelOrder(level: AirbyteLogLevel): number {
  switch (level) {
    case AirbyteLogLevel.FATAL:
      return 60;
    case AirbyteLogLevel.ERROR:
      return 50;
    case AirbyteLogLevel.WARN:
      return 40;
    case AirbyteLogLevel.INFO:
      return 30;
    case AirbyteLogLevel.DEBUG:
      return 20;
    case AirbyteLogLevel.TRACE:
      return 10;
  }
}

export enum AirbyteMessageType {
  CATALOG = 'CATALOG',
  CONNECTION_STATUS = 'CONNECTION_STATUS',
  LOG = 'LOG',
  RECORD = 'RECORD',
  SPEC = 'SPEC',
  STATE = 'STATE',
  TRACE = 'TRACE',
}

export enum AirbyteConnectionStatus {
  SUCCEEDED = 'SUCCEEDED',
  FAILED = 'FAILED',
}

export enum AirbyteTraceFailureType {
  SYSTEM_ERROR = 'system_error',
  CONFIG_ERROR = 'config_error',
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface AirbyteConfig {}

export function parseAirbyteMessage(s: string): AirbyteMessage {
  try {
    const res: AirbyteMessage = JSON.parse(s);
    if (!res.type) {
      throw new VError(`Message type is not set`);
    }
    switch (res.type) {
      case AirbyteMessageType.RECORD:
        return new AirbyteRecord((res as AirbyteRecord).record);
      case AirbyteMessageType.CATALOG:
      case AirbyteMessageType.CONNECTION_STATUS:
      case AirbyteMessageType.LOG:
      case AirbyteMessageType.SPEC:
      case AirbyteMessageType.STATE:
      case AirbyteMessageType.TRACE:
        return res;
      default:
        throw new VError(`Unsupported message type ${res.type}`);
    }
  } catch (e) {
    throw new VError(e, `Invalid Airbyte message: ${s}`);
  }
}

export interface AirbyteMessage {
  readonly type: AirbyteMessageType;
}

export enum SyncMode {
  FULL_REFRESH = 'full_refresh',
  INCREMENTAL = 'incremental',
}

export enum DestinationSyncMode {
  APPEND = 'append',
  OVERWRITE = 'overwrite',
  APPEND_DEDUP = 'append_dedup',
}

export interface AirbyteStream {
  name: string;
  json_schema: Dictionary<any>;
  supported_sync_modes?: SyncMode[];
  source_defined_cursor?: boolean;
  default_cursor_field?: string[];
  source_defined_primary_key?: string[][];
  namespace?: string;
}

export interface AirbyteCatalog {
  streams: AirbyteStream[];
}

export class AirbyteCatalogMessage implements AirbyteMessage {
  readonly type: AirbyteMessageType = AirbyteMessageType.CATALOG;
  constructor(readonly catalog: AirbyteCatalog) {}
}

export interface AirbyteConfiguredStream {
  stream: AirbyteStream;
  sync_mode: SyncMode;
  cursor_field?: string[];
  destination_sync_mode?: DestinationSyncMode;
  primary_key?: string[][];
}

export interface AirbyteConfiguredCatalog {
  streams: AirbyteConfiguredStream[];
}

export class AirbyteConnectionStatusMessage implements AirbyteMessage {
  readonly type: AirbyteMessageType = AirbyteMessageType.CONNECTION_STATUS;
  constructor(
    readonly connectionStatus: {
      status: AirbyteConnectionStatus;
      message?: string;
    }
  ) {}
}

export class AirbyteLog implements AirbyteMessage {
  readonly type: AirbyteMessageType = AirbyteMessageType.LOG;
  constructor(
    readonly log: {
      level: AirbyteLogLevel;
      message: string;
      stack_trace: string;
    }
  ) {}
  static make(
    level: AirbyteLogLevel,
    message: string,
    stack_trace?: string
  ): AirbyteLog {
    return new AirbyteLog({level, message, stack_trace});
  }
}

export const AirbyteRawStreamPrefix = '_airbyte_raw_';
export const AirbyteRawABId = '_airbyte_ab_id';
export const AirbyteRawEmittedAt = '_airbyte_emitted_at';
export const AirbyteRawData = '_airbyte_data';

export class AirbyteRecord implements AirbyteMessage {
  readonly type: AirbyteMessageType = AirbyteMessageType.RECORD;
  constructor(
    readonly record: {
      stream: string;
      namespace?: string;
      emitted_at: number;
      data: Dictionary<any>;
    }
  ) {}

  isRaw(): boolean {
    const stream = this.record.stream;
    return stream && stream.startsWith(AirbyteRawStreamPrefix);
  }

  unpackRaw(): AirbyteRecord {
    if (this.isRaw()) {
      return new AirbyteRecord({
        stream: this.record.stream.slice(AirbyteRawStreamPrefix.length),
        emitted_at: new Date(this.record.data[AirbyteRawEmittedAt]).getTime(),
        data: JSON.parse(this.record.data[AirbyteRawData]),
      });
    }
    return this;
  }

  static make(
    stream: string,
    data: Dictionary<any>,
    namespace?: string
  ): AirbyteRecord {
    return new AirbyteRecord({
      stream,
      namespace,
      emitted_at: Date.now(),
      data,
    });
  }
}

export interface AirbyteTraceError {
  type: 'ERROR';
  emitted_at: number;
  error: {
    message: string;
    internal_message?: string;
    stack_trace?: string;
    failure_type?: AirbyteTraceFailureType;
  };
}

export class AirbyteTrace implements AirbyteMessage {
  readonly type: AirbyteMessageType = AirbyteMessageType.TRACE;
  constructor(
    readonly trace: {
      type: 'ERROR';
      emitted_at: number;
      error: {
        message: string;
        internal_message?: string;
        stack_trace?: string;
        failure_type?: AirbyteTraceFailureType;
      };
    }
  ) {}

  static make(err: any, failure_type?: AirbyteTraceFailureType): AirbyteTrace {
    const wrapped = wrapApiError(err);
    return new AirbyteTrace({
      type: 'ERROR',
      emitted_at: Date.now(),
      error: {
        message: wrapped.message,
        stack_trace: wrapped.stack,
        internal_message: wrapped.name,
        failure_type,
        ...wrapped,
      },
    });
  }
}

export interface Spec {
  documentationUrl?: string;
  changelogUrl?: string;
  connectionSpecification: Dictionary<any>;
  supportsIncremental?: boolean;
  supportsNormalization?: boolean;
  supportsDBT?: boolean;
  supported_destination_sync_modes?: DestinationSyncMode[];
}

export class AirbyteSpec implements AirbyteMessage {
  readonly type: AirbyteMessageType = AirbyteMessageType.SPEC;
  constructor(readonly spec: Spec) {}
}

export interface AirbyteState {
  [stream: string]: any;
}

export class AirbyteStateMessage implements AirbyteMessage {
  readonly type: AirbyteMessageType = AirbyteMessageType.STATE;
  constructor(
    readonly state: {
      data: AirbyteState;
    }
  ) {}
}
