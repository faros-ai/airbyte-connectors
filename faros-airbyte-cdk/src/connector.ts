import {
  AirbyteConfig,
  AirbyteConnectionStatusMessage,
  AirbyteSpec,
} from './protocol';

export abstract class AirbyteConnector {
  /**
   * @returns the spec for this integration. The spec is a JSON-Schema object
   * describing the required configurations (e.g: username and password)
   * required to run this integration.
   */
  abstract spec(): Promise<AirbyteSpec>;

  /**
   * Tests if the input configuration can be used to successfully connect to the
   * integration e.g: if a provided Stripe API token can be used to connect to
   * the Stripe API.
   *
   * @returns Whether the check succeeds or fails with an error message.
   */
  abstract check(
    config: AirbyteConfig
  ): Promise<AirbyteConnectionStatusMessage>;
}
