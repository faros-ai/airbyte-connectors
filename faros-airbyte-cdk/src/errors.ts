import {AxiosError, AxiosRequestConfig, AxiosResponse} from 'axios';
import stream from 'stream';
import {VError, WError} from 'verror';

function formatMessageData(data: any): any {
  return data instanceof stream.Readable ? '[Stream]' : data;
}

function formatRequest(req: AxiosRequestConfig): unknown {
  return {
    baseUrl: req?.baseURL,
    url: req?.url,
    method: req?.method,
    params: req?.params,
  };
}

function formatResponse(res: AxiosResponse): unknown {
  return {
    status: res.status,
    headers: res.headers,
    data: formatMessageData(res.data),
  };
}

function isAxiosError(error: Error): error is AxiosError {
  return (error as any).isAxiosError;
}

/** Strips verbose properties that libraries like Axios attach to errors */
export function wrapApiError(error: Error, message?: string): Error {
  if (!isAxiosError(error)) {
    const cause = VError.cause(error);
    if (cause) {
      // Wrap the cause recursively since it could contain an Axios error
      error = new WError(wrapApiError(cause), error.message);
    }
    return message ? new WError(error, message) : error;
  }

  const prefix = message ? `${message}: ` : '';
  const res = error.response;
  const info = {
    req: error.request
      ? formatRequest({...error.config, ...error.request})
      : formatRequest(error.config),
    res: res ? formatResponse(res) : undefined,
  };
  if (!res) {
    return new VError(
      {info},
      '%sAPI request failed: %s',
      prefix,
      error.message
    );
  }
  const {data, status} = res;
  let msg = `${prefix}API responded with status ${status}`;
  const causeMsg = typeof data == 'string' ? data : data?.error?.message;
  if (causeMsg) {
    msg += `: ${causeMsg}`;
  }
  return new VError({info}, msg);
}
