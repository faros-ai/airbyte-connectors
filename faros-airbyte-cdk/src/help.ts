import {ok} from 'assert';
import _ from 'lodash';
import {upperFirst} from 'lodash';
import {table} from 'table';
import {Dictionary} from 'ts-essentials';
import {VError} from 'verror';

import {
  runBooleanPrompt,
  runNumberPrompt,
  runSelect,
  runStringPrompt,
} from './prompts';

export interface TableRow {
  title: string;
  path: string;
  section: number;
  children?: ReadonlyArray<number>;
  oneOf?: boolean;
  description?: string;
  required: boolean;
  constValue?: string;
  default?: any;
  examples: ReadonlyArray<any>;
  airbyte_secret?: boolean;
  multiline?: boolean;
  type: string;
  items_type?: string;
}

function visitLeaf(
  o: Dictionary<any>,
  curPath: string[],
  section: number,
  required = false
): TableRow {
  const title =
    o.title || curPath.slice(-1)[0].split('_').map(upperFirst).join(' ');
  const leaf = {
    title,
    path: curPath.join('.'),
    section,
    required,
    description: o.description,
    airbyte_secret: o.airbyte_secret,
    default: o.default,
    constValue: o.const,
    multiline: o.multiline,
    examples: o.examples,
    type: o.type,
    items_type: o.items?.type,
  };

  return leaf;
}

export function traverseObject(
  startObject: Dictionary<any>,
  startPath: string[],
  section = 1,
  required = false
): TableRow[] {
  const result: TableRow[] = [];
  // Queue of objects to process in BFS
  const process: [[Dictionary<any>, string[], number, boolean]] = [
    [startObject, startPath, section, required],
  ];
  let newIdx = section + 1;
  while (process.length > 0) {
    const [curObject, curPath, idx, req] = process.shift();
    if (curObject['type'] !== 'object') {
      result.push(visitLeaf(curObject, curPath, idx, req));
      continue;
    }

    ok(curObject.properties || curObject.oneOf);
    ok(curObject.properties === undefined || curObject.oneOf === undefined);
    ok(curObject.title);

    if (curObject.properties) {
      const children = Object.keys(curObject.properties).length;
      if (!children) {
        result.push({
          title: curObject.title,
          path: curPath.join('.'),
          section: idx,
          description: 'Skip this section',
          required: false,
          type: 'empty_object',
          examples: [],
        });
        continue;
      }
      result.push({
        title: curObject.title,
        path: curPath.join('.'),
        section: idx,
        children: _.range(newIdx, newIdx + children),
        oneOf: false,
        description: `Please configure argument${
          children > 1 ? 's' : ''
        } ${_.range(newIdx, newIdx + children).join()} as needed`,
        required: req,
        type: 'object',
        examples: [],
      });
      const requiredProperties: string[] = curObject.required || [];
      const cmp = function compareFn(a: string, b: string): number {
        const aOrder = curObject.properties[a]['order'];
        const bOrder = curObject.properties[b]['order'];
        // Properties with 'order' should appear before than
        // those without.
        if (aOrder !== undefined && bOrder === undefined) return -1;
        if (aOrder === undefined && bOrder !== undefined) return 1;
        // At this point they're both undefined or both
        // with order. Only return if they are not tied.
        if (aOrder !== bOrder) return aOrder - bOrder;
        // Break the tie by comparing the property names.
        return a.localeCompare(b);
      };
      for (const propertyName of Object.keys(curObject.properties).sort(cmp)) {
        process.push([
          curObject.properties[propertyName],
          curPath.concat(propertyName),
          newIdx++,
          requiredProperties.includes(propertyName),
        ]);
      }
    } else {
      const children = curObject.oneOf.length;
      ok(children > 0);
      result.push({
        title: curObject.title,
        path: curPath.join('.'),
        section: idx,
        children: _.range(newIdx, newIdx + children),
        oneOf: true,
        description:
          children > 1
            ? `Please select and configure a single argument from ${_.range(
                newIdx,
                newIdx + children
              ).join()}`
            : `Please configure argument ${newIdx}`,
        required: req,
        type: 'object',
        examples: [],
      });
      for (const choice of curObject.oneOf) {
        process.push([choice, curPath, newIdx++, false]);
      }
    }
  }

  return result;
}

export function helpTable(rows: ReadonlyArray<TableRow>): string {
  const data = [
    [
      'Section ID',
      'Title',
      'Argument',
      'Description',
      'Required',
      'Constant',
      'Default',
      'Examples',
    ],
  ];

  const config = {
    columns: {
      1: {width: 15, wrapWord: true},
      2: {width: 15, wrapWord: true},
      3: {width: 30, wrapWord: true},
      7: {width: 20, wrapWord: true},
    },
  };

  for (const row of rows) {
    data.push([
      row.section,
      row.title,
      row.path,
      row.description || '',
      row.required ? 'Yes' : 'No',
      row.constValue || '',
      row.default ?? '',
      row.examples ? row.examples.join('\n') : '',
    ]);
  }

  return table(data, config);
}

async function promptOneOf(row: TableRow, sections: Map<number, TableRow>) {
  const choices = [];
  if (!row.required) {
    choices.push({
      message: 'Skip this section',
      value: 'Skipped.',
    });
  }
  for (const child of row.children) {
    choices.push({message: sections.get(child).title, value: child});
  }
  const choice = await runSelect({
    name: 'oneOf',
    message: row.title,
    choices,
  });

  if (choice === 'Skipped.') {
    return undefined;
  }

  return +choice;
}

async function promptValue(row: TableRow) {
  const type = row.items_type ?? row.type;
  ok(type);

  const message = row.description
    ? `${row.title}: ${row.description}`
    : row.title;

  switch (type) {
    case 'boolean':
      return await runBooleanPrompt({message});
    case 'integer':
      return await runNumberPrompt({message});
    case 'string':
      return await runStringPrompt({message});
  }

  throw new VError(`Unexpected type: ${type}`);
}

function choiceAsType(row: TableRow, choice: string) {
  const type = row.items_type ?? row.type;
  ok(type);

  switch (type) {
    case 'boolean':
      return choice === 'true';
    case 'integer':
      return +choice;
    case 'string':
      return choice;
  }

  throw new VError(`Unexpected type: ${type}`);
}

function formatArg(row: TableRow, choice: boolean | number | string) {
  let formattedChoice = typeof choice === 'string' ? `"${choice}"` : choice;
  if (row.type === 'array') {
    formattedChoice = `'[${formattedChoice}]'`;
  }
  return `${row.path} ${formattedChoice}`;
}

async function promptLeaf(row: TableRow) {
  if (row.type === 'empty_object') {
    return undefined;
  }
  const choices = [];
  if (row.constValue !== undefined) {
    return row.constValue;
  }
  if (!row.required) {
    choices.push({
      message: 'Skip this section',
      value: 'Skipped.',
    });
  }
  if (row.default !== undefined) {
    choices.push({
      message: `Use default (${row.default})`,
      value: 'Used default.',
    });
  }
  if (row.examples?.length) {
    let idx = 0;
    for (const example of row.examples) {
      idx++;
      choices.push({message: `example ${idx} (${example})`, value: example});
    }
  }

  let choice = ' ';
  if (choices.length) {
    choices.push({
      message: 'Enter your own value',
      value: ' ',
    });
    const message = row.description
      ? `${row.title}: ${row.description}`
      : row.title;
    choice = await runSelect({
      name: 'leaf',
      message,
      choices,
    });
  }

  switch (choice) {
    case 'Skipped.':
      return undefined;
    case 'Used default.':
      return row.default;
    case ' ':
      return await promptValue(row);
    default:
      return choiceAsType(row, choice);
  }
}

export async function buildJson(
  rows: ReadonlyArray<TableRow>
): Promise<string> {
  const result = {};

  await acceptUserInput(rows, (row, choice) =>
    _.set(result, row.path, row.type === 'array' ? [choice] : choice)
  );

  return JSON.stringify(result);
}

export async function buildArgs(
  rows: ReadonlyArray<TableRow>
): Promise<string> {
  const result = [];

  await acceptUserInput(rows, (row, choice) =>
    result.push(formatArg(row, choice))
  );

  return result.join(' \\\n');
}

async function acceptUserInput(
  rows: ReadonlyArray<TableRow>,
  action: (row: TableRow, choice: any) => void
): Promise<void> {
  const sections: Map<number, TableRow> = new Map(
    rows.map((row) => [row.section, row])
  );

  // Stack of sections to process in DFS
  const process = [0];
  const processed = [];
  while (process.length) {
    const section = process.pop();
    processed.push(section);
    const row = sections.get(section);
    if (row.children?.length) {
      if (row.oneOf) {
        const choice = await promptOneOf(row, sections);
        if (choice) {
          process.push(choice);
        }
      } else {
        for (let idx = row.children.length - 1; idx >= 0; idx--) {
          process.push(row.children[idx]);
        }
      }
    } else {
      const choice = await promptLeaf(row);
      if (choice !== undefined) {
        action(row, choice);
      }
    }
  }
}
