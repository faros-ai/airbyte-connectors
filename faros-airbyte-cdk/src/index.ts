export * from './destinations/destination';
export * from './destinations/destination-runner';
export {wrapApiError} from './errors';
export * from './help';
export * from './logger';
export * from './protocol';
export * from './sources/source-base';
export * from './sources/source';
export * from './sources/source-runner';
export * from './sources/streams/stream-base';
export * from './spec-loader';
export * from './utils';
