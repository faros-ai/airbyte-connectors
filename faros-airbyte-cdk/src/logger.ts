import pino, {DestinationStream, Level, Logger} from 'pino';
import stream from 'stream';

import {
  AirbyteLog,
  AirbyteLogLevel,
  AirbyteLogLevelOrder,
  AirbyteMessage,
  AirbyteMessageType,
  AirbyteTrace,
  AirbyteTraceFailureType,
} from './protocol';

export class AirbyteLogger {
  private level = AirbyteLogLevel.INFO;

  constructor(level?: AirbyteLogLevel) {
    if (level) {
      this.level = level;
    } else if (
      process.env.LOG_LEVEL &&
      AirbyteLogLevel[process.env.LOG_LEVEL.toUpperCase()]
    ) {
      this.level = AirbyteLogLevel[process.env.LOG_LEVEL.toUpperCase()];
    }
  }

  error(message: string, stack_trace?: string): void {
    this.write(AirbyteLog.make(AirbyteLogLevel.ERROR, message, stack_trace));
  }

  warn(message: string, stack_trace?: string): void {
    this.write(AirbyteLog.make(AirbyteLogLevel.WARN, message, stack_trace));
  }

  info(message: string, stack_trace?: string): void {
    this.write(AirbyteLog.make(AirbyteLogLevel.INFO, message, stack_trace));
  }

  debug(message: string, stack_trace?: string): void {
    this.write(AirbyteLog.make(AirbyteLogLevel.DEBUG, message, stack_trace));
  }

  trace(message: string, stack_trace?: string): void {
    this.write(AirbyteLog.make(AirbyteLogLevel.TRACE, message, stack_trace));
  }

  traceError(error: any, failure_type?: AirbyteTraceFailureType): void {
    this.write(AirbyteTrace.make(error, failure_type));
  }

  write(msg: AirbyteMessage): void {
    AirbyteLogger.writeMessage(msg, this.level);
  }

  /**
   * Creates a Pino Logger writing messages in Airbyte format
   *
   * @param level logging level
   * @returns Pino Logger
   */
  asPino(level: Level = 'info'): Logger {
    const defaultLevel = AirbyteLogLevel[level.toUpperCase()];

    const destination: DestinationStream = new stream.Writable({
      write: function (chunk, encoding, next): void {
        const msg = JSON.parse(chunk);
        const lvl: AirbyteLogLevel = msg.level
          ? AirbyteLogLevel[pino.levels.labels[msg.level].toUpperCase()]
          : defaultLevel;
        AirbyteLogger.writeMessage(AirbyteLog.make(lvl, msg.msg), defaultLevel);
        next();
      },
    });

    const logger = pino({level}, destination);
    return logger;
  }

  private static writeMessage(
    msg: AirbyteMessage,
    level: AirbyteLogLevel
  ): void {
    if (msg.type === AirbyteMessageType.LOG) {
      const levelOrder = AirbyteLogLevelOrder(level);
      const msgLevelOrder = AirbyteLogLevelOrder((msg as AirbyteLog).log.level);
      if (levelOrder > msgLevelOrder) return;
    }
    console.log(JSON.stringify(msg));
  }
}
