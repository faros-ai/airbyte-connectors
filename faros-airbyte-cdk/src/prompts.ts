import enquirer from 'enquirer';

export interface SelectConfig {
  name: string;
  message: string;
  choices: ReadonlyArray<any>;
}

export function runSelect(cfg: SelectConfig): Promise<string> {
  return new (enquirer as any).Select(cfg).run();
}

export function runBooleanPrompt(cfg: any): Promise<boolean> {
  return new (enquirer as any).BooleanPrompt(cfg).run();
}

export function runNumberPrompt(cfg: any): Promise<number> {
  return new (enquirer as any).NumberPrompt(cfg).run();
}

export function runStringPrompt(cfg: any): Promise<string> {
  return new (enquirer as any).StringPrompt(cfg).run();
}
