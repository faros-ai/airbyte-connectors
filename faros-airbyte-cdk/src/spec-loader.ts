import $RefParser from '@apidevtools/json-schema-ref-parser';
import {VError} from 'verror';

import {AirbyteSpec, Spec} from './protocol';

// Allows to load specs containing references to spec fragments
export class SpecLoader {
  static async loadSpec(specPath: string): Promise<AirbyteSpec> {
    try {
      const dereferenced = (await $RefParser.dereference(specPath)) as Spec;
      return new AirbyteSpec(dereferenced);
    } catch (err: any) {
      throw new VError(
        err,
        `Failed to load spec references for path ${specPath}`
      );
    }
  }
}
