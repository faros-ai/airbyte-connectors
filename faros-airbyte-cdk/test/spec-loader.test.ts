import path from 'path';

import {SpecLoader} from '../src/spec-loader';

const BASE_RESOURCES_DIR = path.join(__dirname, 'resources');

describe('spec loader', () => {
  test('resolves references', async () => {
    expect(
      await SpecLoader.loadSpec(path.join(BASE_RESOURCES_DIR, 'spec.json'))
    ).toMatchSnapshot();
  });

  test('throws when fails to resolve references', async () => {
    expect(() =>
      SpecLoader.loadSpec(path.join(BASE_RESOURCES_DIR, 'spec-bad-ref.json'))
    ).rejects.toThrowError(/^Failed to load spec references/);
  });
});
