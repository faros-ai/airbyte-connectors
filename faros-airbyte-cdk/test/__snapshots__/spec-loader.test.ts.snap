// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spec loader resolves references 1`] = `
AirbyteSpec {
  "spec": Object {
    "connectionSpecification": Object {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": Object {
        "prop1": Object {
          "properties": Object {
            "name": Object {
              "const": "John Doe",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
    "documentationUrl": "https://docs.faros.ai",
  },
  "type": "SPEC",
}
`;
